torch/csrc/jit/python/python_sugared_value.h

```diff
@@ -206,6 +208,13 @@ struct VISIBILITY_HIDDEN ModuleValue : public SugaredValue {
  private:
   Value* self_;
   std::shared_ptr<ConcreteModuleType> concreteType_;
+  // A type hint for this Module that can unlock additional features not
+  // permitted for Modules without hints (e.g. indexing with non-static keys for
+  // ModuleDicts).
+  TypePtr hint_;
+  // If this is a ModuleDict with a hint_, this is a pointer to a dictionary in
+  // the graph that can be reused across lookups.
+  std::shared_ptr<SimpleValue> dict_{nullptr};
```

Comment:
I don't know if we can do this in the presence of writes to the moduledict. I'd be curious if CSE works well here (it might not)

Reply:
I can add a test for that and emit the dict for every lookup if needed.

Reply:
I tried writing to the `ModuleDict` in the body of `forward` using both `[]` and `ModuleDict.update` and neither compiled, so it seems we don't have to worry about this for now? I think it would be a good thing to add to futureproof this work but it's hard to test given that you can't actually update the `ModuleDict` in TorchScript.

Reply:
You can't index into it with `[]` but you can use setAttr. still it's extremely rare, i'm not sure what the best solution is here. One possible solution is to have add a new op that takes in a string and the ModuleDict

Reply:
More i'm thinking about i think we should probably just add a `prim::IndexModuleDict` op or something along those lines

Reply:
What would the schema for the op be? I'd like to create something like the generic list ops, but with one template parameter for the `ModuleDict`'s type (since it's not really a dict), and another for the return type:
```
prim::ModuleDictIndex(t mod_dict, str ind) -> u
```

Reply:
Yea, it would have to be a special op, like IndexTuple